### PR TITLE
Add the ability to copy a lens

### PR DESCRIPTION
**Pull request by adelpopelkova**
Removed `self.parameters`, the only place where it is left is the web app.

### CODE REVIEW

**Review by frenzymadness — APPROVED** [review]

File: lenscalc_web/app.py [hunk 1]
```diff
@@ -48,6 +48,6 @@ def result(variables=None, values=None):
     except ValueError:
         return flask.render_template("index.html", variables=variables_info)
 
-    calculated_values = lens.parameters
+    calculated_values = lens.__dict__
```
Comment (frenzymadness) [inline lenscalc_web/app.py:51]: Be careful here. `__dict__` might contain also extra attributes.

File: lenscalc.py [hunk 2]
```diff
@@ -30,37 +30,20 @@ class Lens:
...
+            value = float(value)
+        object.__setattr__(self, name, value)
 
     def _calculate_replacements(self):
         result = {}
         for variable in self.variables:
-            if self.parameters[variable] is not None:
-                result[variable] = self.parameters[variable]
+            if getattr(self, variable) is not None:
+                result[variable] = getattr(self, variable)
```
Comment (frenzymadness) [inline lenscalc.py:46]:
```suggestion
            if (value := getattr(self, variable)) is not None:
                result[variable] = value
```

File: lenscalc_web/app.py (hunk 1 above)
Comment (adelpopelkova) [reply lenscalc_web/app.py:51]: This is in the web app, so from my understanding no one should be able to add an extra attribute.
[This](https://github.com/adelpopelkova/lenscalc/blob/0f2a6af2cadf17bd3e64badd72470c55e95ddb84/lenscalc_web/templates/result.html#L51) is where the dictionary is used so extra attributes should not affect the result.

**Review by frenzymadness — COMMENTED** [review]
I think it's great. One more thought: The difference between copy and deepcopy is important when an object has some mutable attributes, for example:

```python
class A:
    def __init__(self, x):
        self.x = x

first_object = A(x=5)
second_object = A(x=["foo", "bar", "baz"])
```
For the first object, it does not matter whether you use copy or deepcopy because its attribute is immutable so you don't need to copy x because you cannot change it. However, second_object needs deepcopy to be completely independent because just copy would cause both original and new objects to point to the same mutable list and that can cause issues if you change the content of the list for one object and expect it to stay the same for the other.

Why I'm writing this is the fact that all attributes of Lens is immutable so using a deepcopy might not be necessary. And the tests kind of prove it. So it's not a problem to have the test for deepcopy but you are testing more of the CPython implementation than your own code.

**Review by frenzymadness — APPROVED** [review]

**Pull request merged**